libdmmp/libdmmp_pg.c

```diff
@@ -0,0 +1,175 @@
+			  _dmmp_path_group_free);
+
+
+struct dmmp_path_group *_dmmp_path_group_new(void)
+{
+	struct dmmp_path_group *dmmp_pg = NULL;
+
+	dmmp_pg = (struct dmmp_path_group *)
+		malloc(sizeof(struct dmmp_path_group));
+
+	if (dmmp_pg != NULL) {
+		dmmp_pg->id = 0;
+		dmmp_pg->status = DMMP_PATH_GROUP_STATUS_UNKNOWN;
+		dmmp_pg->priority = 0;
+		dmmp_pg->selector = NULL;
+		dmmp_pg->dmmp_p_count = 0;
+		dmmp_pg->dmmp_ps = NULL;
+	}
+	return dmmp_pg;
+}
+
+int _dmmp_path_group_update(struct dmmp_context *ctx,
+			    struct dmmp_mpath *dmmp_mp,
+			    struct dmmp_path_group *dmmp_pg,
+			    struct pathgroup *pgp)
+{
+	int rc = DMMP_OK;
+	struct path *pp = NULL;
+	struct dmmp_path *dmmp_p = NULL;
+	int i = 0;
+
+	if (VECTOR_SIZE(pgp->paths) == 0) {
+		_error(ctx, "BUG: mpath %s has a empty(no path) path group",
+		       dmmp_mpath_name_get(dmmp_mp));
+		rc = DMMP_ERR_BUG;
+		goto out;
+	}
+
+	pp = VECTOR_LAST_SLOT(pgp->paths);
+	dmmp_pg->id = pp->pgindex & UINT32_MAX;
+	dmmp_pg->status = pgp->status & UINT32_MAX;
+
+	/* ^ TODO(Gris Ge): Assume struct pathgroup does not have invalid
+	 * status, might need to check that.
+	 */
+	if (pgp->priority <= 0)
+		dmmp_pg->priority = DMMP_PATH_GROUP_PRIORITY_UNKNOWN;
+	else
+		dmmp_pg->priority = pgp->priority & UINT32_MAX;
+
+	if (pgp->selector != NULL) {
+		dmmp_pg->selector = strdup(pgp->selector);
+		_dmmp_alloc_null_check(ctx, dmmp_pg->selector, rc, out);
+	} else {
+		_error(ctx, "BUG: Got NULL pgp->selector");
+		rc = DMMP_ERR_BUG;
+		goto out;
+	}
+
+	_debug(ctx, "Got path group %" PRIu32 ", status %" PRIu32
+	       ", priority %" PRIu32 ", selector '%s'",
+	       dmmp_pg->id, dmmp_pg->status, dmmp_pg->priority,
+	       dmmp_pg->selector);
+
+	dmmp_pg->dmmp_ps = (struct dmmp_path **)
+		malloc(sizeof(struct dmmp_path *) * VECTOR_SIZE(pgp->paths));
+	_dmmp_alloc_null_check(ctx, dmmp_pg->dmmp_ps, rc, out);
+	dmmp_pg->dmmp_p_count = VECTOR_SIZE(pgp->paths);
+
+	_debug(ctx, "Got %d path from path group %" PRIu32"",
+	       dmmp_pg->dmmp_p_count, dmmp_pg->id);
+
+	/* Init dmmp_mp->dmmp_pgs */
+	for (i = 0; i < VECTOR_SIZE(pgp->paths); ++i) {
+		dmmp_pg->dmmp_ps[i] = NULL;
+	}
+
+	vector_foreach_slot(pgp->paths, pp, i) {
+		if (pp == NULL) {
+			_warn(ctx, "Got NULL struct path pointer");
+			continue;
+		}
+		dmmp_p = _dmmp_path_new();
+		_dmmp_alloc_null_check(ctx, dmmp_p, rc, out);
+		dmmp_pg->dmmp_ps[i] = dmmp_p;
+
+		rc = _dmmp_path_update(ctx, dmmp_mp, dmmp_pg, dmmp_p, pp);
+		if (rc != DMMP_OK)
+			goto out;
+	}
+
+out:
+	return rc;
+}
+
+void _dmmp_path_group_free(struct dmmp_path_group *dmmp_pg)
+{
+	uint32_t i = 0;
+	if (dmmp_pg == NULL)
+		return;
+
+	if (dmmp_pg->selector != NULL)
+		free((char *) dmmp_pg->selector);
+
+	if (dmmp_pg->dmmp_ps != NULL) {
+		for (i =0; i < dmmp_pg->dmmp_p_count; ++i) {
+			_dmmp_path_free(dmmp_pg->dmmp_ps[i]);
+		}
+		free(dmmp_pg->dmmp_ps);
+	}
+	free(dmmp_pg);
+}
+
+void dmmp_path_array_get(struct dmmp_path_group *mp_pg,
```

Comment:
No. The `dmmp_path_array_get()` functions is accessing properties (`dmmp_ps` and `dmmp_p_count`) of `struct dmmp_path_group` which is defined in `libdmmp_pg.c`, so it has to stay in `libdmmp_pg.c`.